Implemented rough draft of `check` write mode.

davidbarsky opened the pull request:
This is an attempt to address #2536. Of the items to handle in the list, this PR handles: 

- [x] check to the rustfmt bin options
- [x] implement check like diff mode, but with different exit codes.

It does not, at press time: 

- [ ] document the check mode
- [x] add a test to make sure this works. 

I'll add the latter two this weekend.

(This PR is somewhat clunky, as I was able to propagate `WriteMode` by adding an additional field to the `Summary` struct. Alternatives include returning a `FmtResult<(Summary, Option<WriteMode>)>`, but that felt messy.)

davidbarsky commented:
In commit `533df58`, I introduced a simple test case for `--write-mode=check` using [assert-cli](https://docs.rs/assert_cli/), but that introduces a new dependency, and I'm not sure that's a good decision. If introducing this dependency is okay, I'd happily write some additional test cases for the other WriteModes and/or introduce more testing for the CLI options.

Additionally, in commit `b0ece6b5`, I've documented the `check` option in the README.

I'm not too happy with how I'm handling propagation of `WriteMode`—I feel like I'm using the Summary struct as a de-facto context object. Would appreciate advice on handling this better.

nrc commented on src/bin/main.rs:
```diff
@@ -342,6 +346,8 @@ fn main() {
             if summary.has_operational_errors() {
                 1
+            } else if summary.has_diff && summary.write_mode == Some(WriteMode::Check) {
+                1
```
It will probably be easier to change the other exit codes now, rather than later.

davidbarsky replied:
Alrighty! I'll implement the stuff suggested in this comment: https://github.com/rust-lang-nursery/rustfmt/issues/1977#issuecomment-372167773

nrc commented on src/config/options.rs:
```diff
@@ -183,6 +183,9 @@ configuration_option_enum! { WriteMode:
     // Output the changed lines (for internal value only)
     Modified,
+    // Displays how much of the input was processed, but if anything was modified, rustfmt quits
+    // with exit code 0. This option is intended to be run as part of a CI pipeline.
```
I think the exit code is backwards - it should quit with `1` if anything was changed. It displays a diff, if there is one, not progress exactly.

davidbarsky replied:
D'oh—you're right. I'll fix in subsequent commits.

davidbarsky replied:
Fixed/resolved.

nrc commented on src/config/summary.rs:
```diff
@@ -26,6 +28,9 @@ pub struct Summary {
     pub has_diff: bool,
 
+    // What write mode rustfmt was invoked with, if any.
+    pub write_mode: Option<WriteMode>,
```
I don't think this needs to be optional - even if there is no explicit write mode, then there is a default.

davidbarsky replied:
Good point. I'll default it to overwrite?

davidbarsky replied:
Resolved, defaults to Overwrite.

nrc commented on README.md:
```diff
@@ -98,14 +98,16 @@ read data from stdin. Alternatively, you can use `cargo fmt` to format all
 * `diff` Will print a diff between the original files and formatted files to stdout.
          Will also exit with an error code if there are any differences.
+* `check` Similar to `diff`, but is intended to be run in during CI and returns error code 1
+          on failure instead of 4 as `diff` does.  
```
I would document check mode in its own right, rather than comparing to diff.

davidbarsky replied:
Good point; will do.

nrc commented:
Thanks for the PR! The 'hard' bit of running the new write mode looks great. I agree that it would be better to avoid adding the write mode to the summary and have suggested a possible solution in a review comment.

I think adding assert-cli as a dev-dependency is fine. Would be good to test a few of the exit code situations.

davidbarsky commented:
> Thanks for the PR! The 'hard' bit of running the new write mode looks great. I agree that it would be better to avoid adding the write mode to the summary and have suggested a possible solution in a review comment.

Thanks! Will address shortly.

> I think adding assert-cli as a dev-dependency is fine. Would be good to test a few of the exit code situations.

Sounds good, I'll write test cases for those. Should I move the assert-cli tests away from main.rs, or...?

davidbarsky commented:
So far:

- Added better documentation for WriteMode::Check, both in `--help` and README.md.
- Implemented Default for WriteMode, where default is Overwrite. This removes several Options in summary and main.

Still to do:

- Refactor/simplify exit codes (I think the acceptance criteria for this PR should be this.)
- Write assert-cli tests for the new exit code system.

topecongiro commented on src/bin/main.rs:
```diff
@@ -342,6 +346,8 @@ fn main() {
         Ok(summary) => {
             if summary.has_operational_errors() {
                 1
```
This part is a bit wrong. IIRC in check mode, the exit code should be either 0 or 1, but this change does not seem to respect it.

davidbarsky replied:
Thanks for the feedback! I haven't gotten around to implementing https://github.com/rust-lang-nursery/rustfmt/issues/1977#issuecomment-372167773, but when I do later today, this the exit code handling _should_ be far nicer and actually correct.

davidbarsky commented:
Sorry for the delay; work happened. I correct the exit code handling as I understood https://github.com/rust-lang-nursery/rustfmt/issues/1977#issuecomment-372167773. I'd like to clarify—should I log cause with `eprintln!` for non-zero exit codes?

nrc commented:
> should I log cause with eprintln! for non-zero exit codes?

Yes, I think there should be some kind of error message. However, if we're emitting errors or warnings or whatever anyway, then I don't think we need to emit an *additional* message.

nrc commented on src/bin/main.rs:
```diff
@@ -342,13 +346,12 @@ fn main() {
-            } else if summary.has_diff {
-                // should only happen in diff mode
-                4
+                1
```
You could use `||` rather than having multiple `else if` branches here.

davidbarsky replied:
Good point, will update later today.

nrc commented on src/bin/main.rs:
```diff
@@ -287,6 +287,10 @@ fn execute(opts: &Options) -> FmtResult<Summary> {
+            if let Some(mode) = options.write_mode {
+                error_summary.add_write_mode(mode)
+            }
+
```
I thought I suggested a refactoring to avoid saving the write mode in the error summary, but maybe that was on another PR since I can't find any trace of it.

I think that by moving some code around in the main function so that we know the write mode (i.e., process the opts outside of `execute`), then we don't need to save it into the summary.

davidbarsky replied:
> I thought I suggested a refactoring to avoid saving the write mode in the error summary, but maybe that was on another PR since I can't find any trace of it.

No worries.

> I think that by moving some code around in the main function so that we know the write mode (i.e., process the opts outside of execute), then we don't need to save it into the summary.

I think that's a good idea. I wanted to introduce a minimally invasive change with this PR, but if exit codes are gonna change, so should this.

nrc commented on README.md:
```diff
@@ -98,14 +98,16 @@ read data from stdin. Alternatively, you can use `cargo fmt` to format all
 * `plain` Also writes to stdout, but with no metadata.
 * `diff` Will print a diff between the original files and formatted files to stdout.
          Will also exit with an error code if there are any differences.
+* `check` Checks if a diff can be generated. If so, rustfmt outputs a diff and quits with exit code 1.
```
I would re-word, something like: "Checks if the program's formatting matches what rustfmt would do. Silently exits with code 0 if so, emits a diff and exits with code 1 if not. This option is designed to be run in CI-like where a non-zero exit signifies incorrect formatting."

davidbarsky replied:
Reselved.

davidbarsky commented:
Sorry about the slowness on my end. I've been on-call at work, but that's over. With the coming holiday, I should be able to devote some time to landing this issue. Anyways: 

- I updated the README.md as per @nrc's suggestion.
- Made the exit-code handling less bad.
- Moved WriteMode out of the summary, handling parsing in `fn main() {...}`

That being said, the way I'm handling WriteMode isn't great. I think if I'm to refactor how exit code handling and argument parsing are handled, I'd like to use StructOpt, but I'd like to be sure that it's okay I introduce this dependency. Thoughts?

nrc commented:
>  I'd like to use StructOpt, but I'd like to be sure that it's okay I introduce this dependency. Thoughts?

How does using StructOpt help? I'm kind of OK (I like StructOpt), but worry it might be churn for no benefit.

davidbarsky commented:
> How does using StructOpt help? I'm kind of OK (I like StructOpt), but worry it might be churn for no benefit.

I had some more time to think about it—it won't. Oops.

nrc commented:
@davidbarsky How's it going on this? Got any questions I can answer?

davidbarsky commented:
@nrc Hi Nick! Yes! Some guidance around changing `fn execute`. I'm thinking that, based off your suggestion above—quoted for reference:

> I think that by moving some code around in the main function so that we know the write mode (i.e., process the opts outside of execute), then we don't need to save it into the summary.”

—`fn execute` can be broken into a function that handles the parsing of getopts/reading of configuration files, and another function than does the the actual formatting process. The results of the command/config parsing can threaded through `main` so that the result of the attempted run of rustfmt can be associated with the user's command config, thereby not overloading `fn execute` or `Summary` with extraneous information. Does this seem reasonable?

(apologies for the delay.)

nrc commented:
> Does this seem reasonable?

Yes, I think so

davidbarsky commented:
Will need to rebase on mainline, but this I'm getting errors with rustc-ap-rustc_errors. Main thing—I'm not sure how I can make the handling of WriteMode/exit codes better _even_ with refactoring. I'd like to avoid holding up further work/progress on this, I'll post what I've got. If this isn't what y'all are looking for, feel free to take this work and build atop of it/drop it.

davidbarsky commented:
Okay, rebased. Should be conflict-free.

nrc commented:
This looks great, thank you!
